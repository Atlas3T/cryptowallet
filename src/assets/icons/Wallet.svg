<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 22.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 51 51" style="enable-background:new 0 0 51 51;" xml:space="preserve">
<style type="text/css">
	.st0{fill:none;stroke:#000000;stroke-width:5;stroke-miterlimit:10;}
	.st1{display:none;}
	.st2{display:inline;fill:none;stroke:#000000;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:10;}
	.st3{display:inline;}
</style>
<g id="Wallet">
	<path class="st0" d="M46.1,31.5v5.3c0,2.4-1.9,4.3-4.3,4.3H9.5c-2.4,0-4.3-1.9-4.3-4.3V15.5c0-2.4,1.9-4.3,4.3-4.3h32.3
		c2.4,0,4.3,1.9,4.3,4.3v5.2"/>
	<circle cx="46.1" cy="26.1" r="3.1"/>
</g>
<g id="Payments" class="st1">
	<line class="st2" x1="8" y1="19.8" x2="41.8" y2="19.8"/>
	<polyline class="st2" points="35.8,13.3 42.3,19.8 35.8,26.3 	"/>
	<line class="st2" x1="42.3" y1="32.8" x2="8.5" y2="32.8"/>
	<polyline class="st2" points="14.5,39.3 8,32.8 14.5,26.3 	"/>
</g>
<g id="Exchange" class="st1">
	<path class="st2" d="M15.3,21.7c0-5.6,4.6-10.2,10.2-10.2s10.2,4.6,10.2,10.2"/>
	<polyline class="st2" points="42.2,16.3 35.7,22.8 29.2,16.3 	"/>
	<path class="st2" d="M35.7,30.2c0,5.6-4.6,10.2-10.2,10.2s-10.2-4.6-10.2-10.2"/>
	<polyline class="st2" points="8.8,35.7 15.3,29.2 21.9,35.7 	"/>
</g>
<g id="More" class="st1">
	<circle class="st3" cx="26" cy="15.8" r="7.8"/>
	<circle class="st3" cx="12.3" cy="28.6" r="5"/>
	<circle class="st3" cx="35.3" cy="33.6" r="3.5"/>
</g>
</svg>
