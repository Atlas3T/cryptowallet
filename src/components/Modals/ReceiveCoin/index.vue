<template>
  <div>
    <q-modal
      v-model="receiveCoinModalOpened"
      class="light-modal modal"
    >
      <div class="header-section">
        <div class="header-back-button-wrapper">
          <q-btn
            icon="arrow_back"
            size="lg"
            class="icon-btn back-arrow-btn"
            flat
            @click.prevent="goBack"
          />
        </div>
        <h1 class="header-h1">Receive</h1>
      </div>
      <div
        v-if="wallet"
        class="modal-layout-wrapper"
      >
        <Receive/>
      </div>
    </q-modal>
  </div>
</template>

<script>
import { mapState } from 'vuex';
import Receive from '@/components/Wallet/ReceiveCoin';

export default {
  name: 'ReceiveCoin',
  components: {
    Receive,
  },
  data() {
    return {
      receiveCoinModalOpened: false,
    };
  },
  computed: {
    ...mapState({
      id: state => state.route.params.id,
      authenticatedAccount: state => state.settings.authenticatedAccount,
    }),
    wallet() {
      return this.$store.getters['entities/wallet/find'](this.id);
    },
  },
  watch: {
    $route: {
      handler(to) {
        if (to.name === 'receiveCoin') {
          this.receiveCoinModalOpened = true;
        } else {
          this.receiveCoinModalOpened = false;
        }
      },
    },
    receiveCoinModalOpened: {
      handler(newVal, oldVal) {
        if (oldVal === true && newVal === false) {
          if (this.$store.state.route.name === 'receiveCoin') this.$router.go(-1);
        }
      },
    },
  },
  mounted() {
    this.$root.$on('receiveCoinModalOpened', (value) => {
      this.receiveCoinModalOpened = value;
    });
  },
  methods: {
    goBack() {
      this.$router.go(-1);
    },
  },
};
</script>

<style>
</style>
