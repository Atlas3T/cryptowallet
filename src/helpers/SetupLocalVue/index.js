import Vuex from 'vuex';
import { createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import VueI18n from 'vue-i18n';
import messages from '@/i18n/';
import axios from 'axios';
import { sync } from 'vuex-router-sync';
import errorHandlerPlugin from '@/boot/ErrorHandler';
import toasterPlugin from '@/boot/Toaster';
import permissionsPlugin from '@/boot/Permissions';
import appInvitationPlugin from '@/boot/AppInvitation';
import smsPlugin from '@/boot/Sms';
import emailPlugin from '@/boot/Email';
import contactsImportPlugin from '@/boot/ContactsImport';
import * as All from 'quasar'

const { Quasar, date } = All;

const components = Object.keys(All).reduce((object, key) => {
  const val = All[key]
  if (val && val.component && val.component.name != null) {
    object[key] = val
  }
  return object
}, {})


const localVue = createLocalVue();

localVue.use(Quasar, { components });
localVue.use(VueI18n);
localVue.use(Vuex);
// localVue.use(VueRouter);
localVue.prototype.$axios = axios;
errorHandlerPlugin({ Vue: localVue });
toasterPlugin({ Vue: localVue });
permissionsPlugin({ Vue: localVue });
appInvitationPlugin({ Vue: localVue });
smsPlugin({ Vue: localVue });
emailPlugin({ Vue: localVue });
contactsImportPlugin({ Vue: localVue });

/**
 * creates a new router and syncs it with the store using vuex-router-sync
 * @param store
 * @returns router
 */
const createRouter = (store = null) => {
  const router = new VueRouter({
    routes: [
      {
        path: '/contact/:id',
      },
    ],
  });

  if (store) { sync(store, router); }
  return router;
};

const i18n = new VueI18n({ messages, locale: 'en-gb' });

export {
  localVue,
  createRouter,
  i18n,
};
