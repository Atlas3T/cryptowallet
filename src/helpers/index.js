import AmountFormatter from './AmountFormatter';
import dateTranslater from './DateTranslater';
import Contacts from './Contacts';

export {
  AmountFormatter,
  dateTranslater,
  Contacts,
};
