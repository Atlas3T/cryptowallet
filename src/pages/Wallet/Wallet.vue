<template>
  <div class="layout">
    <div class="list">
      <WalletsList
        :coins="coins"
        :currency="currency"
      />
    </div>
    <div class="button">
      <AddWallet/>
    </div>
  </div>
</template>

<script>
import WalletsList from '@/components/Wallet/Wallet/WalletsList.vue';
import AddWallet from '@/components/Wallet/Wallet/AddWallet.vue';

export default {
  name: 'Balance',
  components: {
    WalletsList,
    AddWallet,
  },
  /**
   * @TODO James Konrad
   * data below is mocked
   *
   * currency String is a prefered user currency, should be in the Vuex store
   * you should probably select your prefered currency during the initial setup
   * alongside the language
   *
   * coins Object also shoudn't be here
   */
  data() {
    return {
      currency: 'GPB',
      coins: {
        BTC: {
          key: 'BTC',
          name: 'Bitcoin',
          balance: 25.532344,
          uid: 1,
        },
        ATL: {
          key: 'ATL',
          name: 'Atlas',
          balance: 21.532344,
          uid: 2,
        },
        ETH: {
          key: 'ETH',
          name: 'Ethereum',
          balance: 55.532344,
          uid: 3,
        },
      },
    };
  },
};

</script>

<style scoped>
.layout {
  display: flex;
  flex-direction: column;
}

.list {
  flex: 1;
  position: relative;
}
</style>
