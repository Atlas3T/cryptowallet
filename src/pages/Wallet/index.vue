<template>
  <div class="layout">
    <div class="list">
      <wallet-list
        :wallets="wallets"
        :currency="selectedCurrency"
      />
    </div>
    <div class="button">
      <add-wallet/>
    </div>
  </div>
</template>

<script>
import WalletList from '@/components/Wallet/WalletsList';
import AddWallet from '@/components/Wallet/AddWallet';

export default {
  name: 'Balance',

  components: {
    WalletList,
    AddWallet,
  },

  /**
   * @TODO James Konrad
   * data below is mocked
   *
   * currency String is a prefered user currency, should be in the Vuex store
   * you should probably select your prefered currency during the initial setup
   * alongside the language
   *
   * coins Object also shoudn't be here
   */
  data() {
    return {
      wallets: {
        BTC: {
          key: 'BTC',
          name: 'Bitcoin',
          balance: 25.532344,
          uid: 1,
        },
        ATL: {
          key: 'ATL',
          name: 'Atlas',
          balance: 21.532344,
          uid: 2,
        },
        ETH: {
          key: 'ETH',
          name: 'Ethereum',
          balance: 55.532344,
          uid: 3,
        },
      },
    };
  },

  computed: {
    selectedCurrency() {
      return this.$store.state.settings.selectedCurrency;
    },
  },
};

</script>

<style scoped>
.layout {
  display: flex;
  flex-direction: column;
}

.list {
  flex: 1;
  position: relative;
}
</style>
