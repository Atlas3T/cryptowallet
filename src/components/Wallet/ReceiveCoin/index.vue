<template>
  <div class="receive-coin-box">
    <CoinHeader :wallet="wallet" />

    <div class="send-modal-heading">
      <h3>{{ $t('yourAddress') }}</h3>
      <span class="h3-line" />
    </div>
    <div class="address break">
      {{ address }}
    </div>

    <q-btn
      :label="$t('copy')"
      color="blueish"
      size="sm"
      @click="copyToClipboard"
    />

    <q-btn
      :label="$t('share')"
      class="share-btn"
      color="blueish"
      size="sm"
      @click="share()"
    />

    <div class="send-modal-heading">
      <h3>{{ $t('scanQR') }}</h3>
      <span class="h3-line" />
    </div>

    <div class="qr-code">
      <img
        v-if="qrCodeDataURL"
        :src="qrCodeDataURL"
      >
    </div>
  </div>
</template>

<script>
import { mapState } from 'vuex';
import QRCode from 'qrcode';
import CoinHeader from '@/components/Wallet/CoinHeader';
import Wallet from '@/store/wallet/entities/wallet';
import Address from '@/store/wallet/entities/address';

export default {
  name: 'Receive',
  components: {
    CoinHeader,
  },
  data() {
    return {
      qrCodeDataURL: null,
    };
  },
  computed: {
    ...mapState({
      id: (state) => { return state.route.params.id; },
      authenticatedAccount: (state) => { return state.settings.authenticatedAccount; },
      delay: (state) => { return state.settings.delay; },
    }),
    wallet() {
      return this.$store.getters['entities/wallet/find'](this.id);
    },
    address() {
      if (this.wallet.externalAddress) { return this.wallet.externalAddress; }
      return this.generateExternalAddress();
    },
  },
  watch: {
    address() {
      console.log(this.address);
      this.qrCode();
    },
  },
  mounted() {
    this.qrCode();
  },
  methods: {
    generateExternalAddress() {
      let address;
      if (this.wallet.sdk === 'ERC20') {
        ({ address } = this.wallet.erc20Wallet);
      } else {
        const coinSDK = this.coinSDKS[this.wallet.sdk];

        const keyPair = coinSDK.generateKeyPair(
          this.wallet.hdWallet, this.wallet.externalChainAddressIndex,
        );

        const newAddress = {
          account_id: this.authenticatedAccount,
          wallet_id: this.wallet.id,
          chain: 'external',
          address: keyPair.address,
          index: this.wallet.externalChainAddressIndex,
        };

        Address.$insert({ data: newAddress });

        ({ address } = keyPair);
      }
      if (this.wallet.sdk === 'Bitcoin') {
        Wallet.$update({
          where: (record) => { return record.id === this.wallet.id; },
          data: {
            externalChainAddressIndex: this.wallet.externalChainAddressIndex + 1,
            externalAddress: address,
          },
        });
      } else {
        Wallet.$update({
          where: (record) => { return record.id === this.wallet.id; },
          data: {
            externalChainAddressIndex: 0,
            externalAddress: address,
          },
        });
      }

      return address;
    },
    copyToClipboard() {
      try {
        cordova.plugins.clipboard.copy(this.address);
        this.$toast.create(0, this.$t('copied'), this.delay.short);
      } catch (err) {
        this.errorHandler(err);
      }
    },
    qrCode() {
      const options = {
        width: 250,
        height: 250,
      };

      if (typeof this.address !== 'string') { return false; }
      QRCode.toDataURL(this.address, options, (err, url) => {
        if (err) {
          this.errorHandler(err);
          return false;
        }
        this.qrCodeDataURL = url;
        return false;
      });

      return false;
    },
    share() {
      const options = {
        message: `${this.address}`,
        subject: `Here's my ${this.wallet.name} address`,
      };

      const onError = (msg) => {
        this.errorHandler(new Error(msg));
      };
      window.plugins.socialsharing.shareWithOptions(options, () => {}, onError);
    },
  },
};
</script>

<style scoped>

.address {
  font-size: 0.8rem;
  margin: 1rem 0;
}

.break {
  word-break: break-all;
}

.to-copy-input {
  position: absolute;
  left: -9999px;
}

.qr-code {
  margin: 0rem;
  display: flex;
  justify-content: center;
}

.share-btn {
  margin-left: 0.5rem;
}
</style>
