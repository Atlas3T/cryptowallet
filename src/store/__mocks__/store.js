import Vue from 'vue';
import Vuex from 'vuex';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from '@atlascity/vuex-orm-plugin-lokijs';

import wallet from '@/store/wallet';
import contacts from '@/store/contacts/__mocks__/contacts.js';
import search from '@/store/search/__mocks__/search.js';
import settings from '@/store/settings/__mocks__/settings.js';
import setup from '@/store/setup/__mocks__/setup.js';
import qrcode from '@/store/qrcode/__mocks__/qrcode.js';
import modals from '@/store/modals/__mocks__/modals.js';

// entities
import Account from '@/store/wallet/entities/account';
import Address from '@/store/wallet/entities/address';
import Tx from '@/store/wallet/entities/tx';
import Utxo from '@/store/wallet/entities/utxo';
import Wallet from '@/store/wallet/entities/wallet';
import Coin from '@/store/wallet/entities/coin';
import KeyPair from '@/store/wallet/entities/keyPair';
import Prices from '@/store/prices';
import LatestPrice from '@/store/latestPrice';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
  Wallet,
];

const options = {
  env: 'browser',
};

VuexORM.use(VuexORMLoki, { database, options });

export const actions = {};
export const getters = {};
export const mutations = {};
export const state = {};

export function createMocks(custom = {
  getters: {}, mutations: {}, actions: {}, state: {},
}) {
  /**
   * mock store modules.
   */
  const modules = {
    wallet,
    search,
    settings,
    contacts,
    setup,
    qrcode,
    modals,
  };

  let mockGettersCombined = {};
  let mockMutationsCombined = {};
  let mockActionsCombined = {};
  let mockStateCombined = {};

  Object.entries(modules).forEach(([name, module]) => {
    mockGettersCombined = Object.assign({}, mockGettersCombined, module.getters);
    mockMutationsCombined = Object.assign({}, mockMutationsCombined, module.mutations);
    mockActionsCombined = Object.assign({}, mockActionsCombined, module.actions);
    mockStateCombined = Object.assign({}, mockStateCombined, module.state);

    if (custom.state && custom.state[name]) {
      module.state = Object.assign(module.state, custom.state[name]);
    }

    modules[name] = {
      namespaced: true,
      getters: Object.assign({}, getters, module.getters),
      mutations: Object.assign({}, mutations, module.mutations),
      actions: Object.assign({}, actions, module.actions),
      state: Object.assign({}, state, module.state),
    };
  });

  mockGettersCombined = Object.assign({}, mockGettersCombined, custom.getters);
  mockMutationsCombined = Object.assign({}, mockMutationsCombined, custom.mutations);
  mockActionsCombined = Object.assign({}, mockActionsCombined, custom.actions);
  mockStateCombined = Object.assign({}, mockStateCombined, custom.state);
  return {
    getters: mockGettersCombined,
    mutations: mockMutationsCombined,
    actions: mockActionsCombined,
    state: mockStateCombined,
    store: new Vuex.Store({
      getters: mockGettersCombined,
      mutations: mockMutationsCombined,
      actions: mockActionsCombined,
      state: mockStateCombined,
      modules,
      plugins: [VuexORM.install(database)],
    }),
  };
}

export const { store } = createMocks();
