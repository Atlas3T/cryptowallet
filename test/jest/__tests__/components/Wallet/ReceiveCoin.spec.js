/* eslint-disable no-magic-numbers */
import ReceiveCoin from '@/components/Wallet/ReceiveCoin';
import { mount } from '@vue/test-utils';
import { localVue, i18n, createRouter } from '@/helpers/SetupLocalVue';
import { createMocks as createStoreMocks } from '@/store/__mocks__/store.js';
import Wallet from '@/store/wallet/entities/wallet';
import Address from '@/store/wallet/entities/address';
import cordovaMocks from '~/test/CordovaMocks';
import QRCode from 'qrcode';


describe('ReceiveCoin component', () => {
  let wrapper;
  let router;
  let storeMocks;

  const walletData = JSON.parse('{"$id":3,"id":3,"account_id":1,"name":"Bitcoin","displayName":"Bitcoin","symbol":"BTC","sdk":"Bitcoin","network":"BITCOIN_TESTNET","internalChainAddressIndex":23,"externalChainAddressIndex":15,"externalAddress": null,"confirmedBalance":0.52543197,"unconfirmedBalance":0,"imported":false,"enabled":false,"lastBlockHeight":0,"parentName":"","parentSdk":"","contractAddress":"","decimals":"","hdWallet":{"bip":49,"ext":{"xpriv":"xprv9ztbaJYMADTpxd5o1wyCqJAfnBjbLvSqWsqgk2HKSgDsi4BQPBf5PdJ4XLPZd4ugHQ9o45sExCuJuWBnxyu61Gt7fVj1kAPoB1hsze1XNAC","xpub":"xpub6Dswyp5Ezb28B7AG7yWDCS7QLDa5kPAgt6mHYQgw11krarWYviyKwRcYNbeutQRsNtQn1DD9SnaXHMpwpzAEDnBxvZZrDFQ2ehUGw8kxH6u"},"int":{"xpriv":"xprv9ztbaJYMADTpzuEkRPVdJkHNUmdBg98S2VUv91asoQzVTZtHsavn9SYuxy4iYwjpLZrmNy4oUJQoS6fGUJBMvT7Kfm3uvw1Q9b3DhJY23i5","xpub":"xpub6Dswyp5Ezb28DPKDXR2dftE72oTg5brHPiQWwPzVMkXULNDSR8F2hEsPpEZKhymsboomxhqiwmnvec5FTc36WuFWSGPtjQrL6YrQ6JWSBnh"},"type":1,"network":{"name":"BITCOIN_TESTNET","type":"testnet","bip":1,"segwit":true,"discovery":"http://92.207.178.198:3001/api","broadcastUrl":"https://chain.so/api/v2/send_tx/BTCTEST","feeApi":"https://api.blockcypher.com/v1/btc/main","connect":{"messagePrefix":"0018Bitcoin Signed Message:","bech32":"tb","bip32":{"public":70617039,"private":70615956},"pubKeyHash":111,"scriptHash":196,"wif":239}}},"erc20Wallet":""}');
  const addressData = JSON.parse('{"$id":1,"id":1,"wallet_id":3,"account_id":1,"address":"2MwSB1utt5aMRp8tY92wNjBpb96UfpDKHX7","index":0,"chain":"external","used":false}');
  const erc20WalletData = JSON.parse('{"$id":45,"id":45,"account_id":10,"name":"Catalyst","displayName":"Catalyst","symbol":"CAT","sdk":"ERC20","network":"ETHEREUM_ROPSTEN","internalChainAddressIndex":0,"externalChainAddressIndex":0,"externalAddress":null,"confirmedBalance":0,"unconfirmedBalance":0,"imported":false,"enabled":false,"lastBlockHeight":0,"parentName":"Ethereum","parentSdk":"Ethereum","contractAddress":"0x26705403968a8c73656a2fed0f89245698718f3f","decimals":3,"hdWallet":"","erc20Wallet":{"decimals":3,"address":"0xC25Ec706b3835B71Ec713dDAD3557c8413b76913","network":{"name":"ETHEREUM_ROPSTEN","networkName":"ropsten","bip":60,"getTranApi":"http://api-ropsten.etherscan.io/api?module=account&action=txlist&address=","getBalanceApi":"https://api-ropsten.etherscan.io/api?module=account&action=balance&address=","getErc20TranApi":"http://api-ropsten.etherscan.io/api?module=account&action=tokentx&contractaddress=","feeApi":"https://api.blockcypher.com/v1/eth/main","provider":"https://ropsten.infura.io/v3/352fc30cd8364caabaea4a3d67da773f","chainId":3},"name":"Catalyst","symbol":"CAT","contract":"0x26705403968a8c73656a2fed0f89245698718f3f"}}');
  const coinSDKSMock = {
    Bitcoin: {
      generateKeyPair: jest.fn().mockReturnValue({ address: '2NGBz7mknbB1GxFSddxa47C3S6qS4FuTnyd' }),
    },
  };


  const defaultProps = {};
  function wrapperInit(options) {
    return mount(ReceiveCoin, options);
  }

  function storeInit(custom, propsData, path = '/wallet/single/receive/3') {
    storeMocks = createStoreMocks(custom);
    Wallet.insert({ data: walletData, erc20WalletData });
    Address.insert({ data: addressData });

    router = createRouter(storeMocks.store);
    router.push({ path });
    wrapper = wrapperInit({
      i18n,
      router,
      localVue,
      propsData,
      store: storeMocks.store,
      mocks: {
        coinSDKS: coinSDKSMock,
        errorHandler: jest.fn(),
      },
      stubs: {
        CoinHeader: true,
      },
    });
  }

  beforeEach(() => {
    cordovaMocks.initMocks();
    storeInit({}, defaultProps);
  });

  afterEach(() => { cordovaMocks.destroyMocks(); });

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('qrCode()', () => {
    it('it renders a qr code', async (done) => {
      setTimeout(() => {
        expect(wrapper.contains('img')).toBe(true);
        done();
      }, 1000);
      QRCode.toDataURL = jest.fn().mockImplementationOnce((address, options, onError) => {
        return onError('error', {});
      });
    });

    it('can handle errors when the copy button is clicked', async (done) => {
      setTimeout(() => {
        expect(wrapper.vm.errorHandler).toHaveBeenCalled();
        done();
      }, 1000);
    });
  });

  describe('copyToClipboard()', () => {
    it('can copy the address when the copy button is clicked', async (done) => {
      storeInit({}, defaultProps, '/wallet/single/receive/4');
      wrapper.vm.$toast.create = jest.fn();
      cordova.plugins.clipboard.mockBehaviour = 1;
      wrapper.findAll('button').at(0).trigger('click');
      setTimeout(() => {
        expect(wrapper.vm.$toast.create).toHaveBeenCalled();
        done();
      }, 1000);
    });

    it('can handle errors if a qr code cannot be generated', async (done) => {
      wrapper.vm.$toast.create = jest.fn();
      cordova.plugins.clipboard.mockBehaviour = 2;
      wrapper.findAll('button').at(0).trigger('click');
      setTimeout(() => {
        expect(wrapper.vm.errorHandler).toHaveBeenCalled();
        done();
      }, 1000);
    });
  });

  describe('share()', () => {
    it('calls the share api if the share button is clicked', async (done) => {
      const mockPlugins = {
        socialsharing: {
          shareWithOptions: jest.fn().mockImplementationOnce((options, onSuccess) => {
            return onSuccess();
          }),
        },
      };
      global.plugins = mockPlugins;
      wrapper.findAll('button').at(1).trigger('click');
      setTimeout(() => {
        expect(mockPlugins.socialsharing.shareWithOptions).toHaveBeenCalledTimes(1);
        done();
      }, 1000);
    });

    it('can handle errors when the share button is clicked', async (done) => {
      const mockPlugins = {
        socialsharing: {
          shareWithOptions: jest.fn().mockImplementationOnce((options, onSuccess, onError) => {
            return onError('Test Error');
          }),
        },
      };
      global.plugins = mockPlugins;
      wrapper.findAll('button').at(1).trigger('click');
      setTimeout(() => {
        expect(mockPlugins.socialsharing.shareWithOptions).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.errorHandler).toHaveBeenCalledWith(Error('Test Error'));
        done();
      }, 1000);
    });
  });
});
