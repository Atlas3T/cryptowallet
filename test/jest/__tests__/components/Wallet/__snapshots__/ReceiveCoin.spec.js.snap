// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReceiveCoin component renders and matches snapshot 1`] = `
<div
  class="receive-coin-box"
>
  <coinheader-stub
    wallet="[object Object]"
  />
   
  <div
    class="send-modal-heading"
  >
    <h3>
      Your Address
    </h3>
     
    <span
      class="h3-line"
    />
  </div>
   
  <div
    class="address break"
  >
    
    2NGBz7mknbB1GxFSddxa47C3S6qS4FuTnyd
  
  </div>
   
  <button
    class="q-btn inline q-btn-item non-selectable q-btn--standard q-btn--rectangle bg-blueish text-white q-focusable q-hoverable"
    style="font-size: 10px;"
    tabindex="0"
    type="button"
  >
    <div
      class="q-focus-helper"
      tabindex="-1"
    />
    <div
      class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
    >
      <div>
        Copy
      </div>
    </div>
  </button>
   
  <button
    class="q-btn inline q-btn-item non-selectable share-btn q-btn--standard q-btn--rectangle bg-blueish text-white q-focusable q-hoverable"
    style="font-size: 10px;"
    tabindex="0"
    type="button"
  >
    <div
      class="q-focus-helper"
      tabindex="-1"
    />
    <div
      class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
    >
      <div>
        Share
      </div>
    </div>
  </button>
   
  <div
    class="send-modal-heading"
  >
    <h3>
      Scan QR Code
    </h3>
     
    <span
      class="h3-line"
    />
  </div>
   
  <div
    class="qr-code"
  >
    <!---->
  </div>
</div>
`;
